src/R2N.jl

```diff
@@ -1,297 +1,502 @@
+
+# Output
+The value returned is a `GenericExecutionStats`, see `SolverCore.jl`.
+
+# Callback
+The callback is called at each iteration.
+The expected signature of the callback is `callback(nlp, solver, stats)`, and its output is ignored.
+Changing any of the input arguments will affect the subsequent iterations.
+In particular, setting `stats.status = :user` will stop the algorithm.
+All relevant information should be available in `nlp` and `solver`.
+Notably, you can access, and modify, the following:
+- `solver.xk`: current iterate;
+- `solver.∇fk`: current gradient;
+- `stats`: structure holding the output of the algorithm (`GenericExecutionStats`), which contains, among other things:
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.solver_specific[:smooth_obj]`: current value of the smooth part of the objective function
+  - `stats.solver_specific[:nonsmooth_obj]`: current value of the nonsmooth part of the objective function
+  - `stats.status`: current status of the algorithm. Should be `:unknown` unless the algorithm has attained a stopping criterion. Changing this to anything other than `:unknown` will stop the algorithm, but you should use `:user` to properly indicate the intention.
+  - `stats.elapsed_time`: elapsed time in seconds.
 """
 function R2N(
-  f::AbstractNLPModel,
-  h::H,
-  options::ROSolverOptions{R};
-  x0::AbstractVector = f.meta.x0,
-  subsolver_logger::Logging.AbstractLogger = Logging.NullLogger(),
-  subsolver = R2,
-  subsolver_options = ROSolverOptions(ϵa = options.ϵa),
-  Mmonotone::Int = 1,
-  selected::AbstractVector{<:Integer} = 1:(f.meta.nvar),
-) where {H, R}
-  start_time = time()
-  elapsed_time = 0.0
-  # initialize passed options
-  ϵ = options.ϵa
-  ϵ_subsolver_init = subsolver_options.ϵa
-  ϵ_subsolver = copy(ϵ_subsolver_init)
-  ϵr = options.ϵr
-  Δk = options.Δk
-  verbose = options.verbose
-  maxIter = options.maxIter
-  maxTime = options.maxTime
-  η1 = options.η1
-  η2 = options.η2
-  γ = options.γ
-  θ = options.θ
-  σmin = options.σmin
-  α = options.α
-  β = options.β
-  σk = options.σk
-
-  # store initial values of the subsolver_options fields that will be modified
-  ν_subsolver = subsolver_options.ν
-  ϵa_subsolver = subsolver_options.ϵa
-
-  local l_bound, u_bound
-  if has_bounds(f)
-    l_bound = f.meta.lvar
-    u_bound = f.meta.uvar
-  end
+  nlp::AbstractNLPModel{T, V},
+  h,
+  options::ROSolverOptions{T};
+  kwargs...,
+) where {T <: Real, V}
+  kwargs_dict = Dict(kwargs...)
+  selected = pop!(kwargs_dict, :selected, 1:(nlp.meta.nvar))
+  x0 = pop!(kwargs_dict, :x0, nlp.meta.x0)
+  reg_nlp = RegularizedNLPModel(nlp, h, selected)
+  return R2N(
+    reg_nlp,
+    x = x0,
+    atol = options.ϵa,
+    rtol = options.ϵr,
+    neg_tol = options.neg_tol,
+    verbose = options.verbose,
+    max_iter = options.maxIter,
+    max_time = options.maxTime,
```

Comment:
@MaxenceGollier We should also initialize $\sigma_k$ for R2N, as done for R2DH https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl/blob/master/src/R2N.jl#L75
Since ```Bk``` is generally of order 1, this helps preserve the quality of the initial initialization.